### PR TITLE
WIP Make Atom.t private (and consequences)

### PR DESCRIPTION
**Pull request by Chris00**
This is to discuss of making `Atom.t` private (see also https://github.com/ocaml/dune/pull/459).

### CODE REVIEW

**Comment by Chris00** [comment]
The problem is indeed what quantities must be represented as `Atom.t` and where the conversion from `string` to `Atom.t` must take place.

**Comment by Chris00** [comment]
I wonder whether to add a module with various kind of behaviors.  `Atom.t_or_empty` would for example be useful at a few places...

**Comment by Chris00** [comment]
Also, maybe the name `Atom` is not so nice to pervasively use in the code.  Maybe a module `ASCII` or so would be more meaningful?

File: src/usexp/usexp.mli [hunk 1]
```diff
@@ -1,17 +1,43 @@
 (** Parsing of s-expressions *)
 
 module Atom : sig
-  type t = string
+  type t = private A of string [@@unboxed]
```
Comment (diml) [inline src/usexp/usexp.mli]: Nice! That's much better than a private alias

**Comment by diml** [comment]
I think that we need to convert to/from Atom.t in to_sexp/of_sexp functions only. Being an atom or quoted_string is a property of the sexp representation so it shouldn't appear in internal datatypes

**Comment by Chris00** [comment]
> I think that we need to convert to/from Atom.t in to_sexp/of_sexp functions only.

We could but I tend to disagree.  Why have an invariant just to abandoned it minutes later.  This invariant is nice to express what module names are for example.  I agree it should bear a more meaningful name in the rest of the code and, in trying to spread it, I found difficult to do it in a _principled_ way but that's because I do not understand well the whole project.   Still, adding a restricted type of string and in turn precisely defining where the checks occur would strengthen the project IMHO.

**Comment by Chris00** [comment]
Pushed the consequences—not complete—of trying to keep the restricted string invariant to https://github.com/Chris00/dune/tree/private_atom_full

**Comment by Chris00** [comment]
I pushed minimal changes making `Atom.t` private.

**Comment by diml** [comment]
I think it'd be nice to have more private types for the various string types, such as `Package.Name.t`, `Module.Name.t`, etc... However, it's not clear to me why they should be `Atom.t`. For instance, if we supported several formats instead of just sexp for instance, why would we choose `Atom.t` rather than something else?

Moreover, it's not clear that the invariants for atoms are the right ones for everything. For instance, why should we forbid to put parentheses in package names? Even though it doesn't seem like a good idea, it's not forbidden by opam.

Looking again at Sexp.Combinators, actually I find it a bit weird that it mentions `atom` at all. These names are supposed to match OCaml types. I.e. you should use `Sexp.To_sexp.int` to serialize an `int` and you should use `Sexp.To_sexp.string` to serialize a string, unless it is important that it is serialized as a quoted_string.

File: src/action.ml [hunk 2]
```diff
@@ -69,38 +69,45 @@ struct
...
+                                string k ; string v ; sexp_of_t r]
     | Redirect (outputs, fn, r) ->
-      List [ Atom (sprintf "with-%s-to" (Outputs.to_string outputs))
+      List [ Quoted_string (sprintf "with-%s-to" (Outputs.to_string outputs))
            ; path fn
            ; sexp_of_t r
            ]
     | Ignore (outputs, r) ->
-      List [ Atom (sprintf "ignore-%s" (Outputs.to_string outputs))
+      List [ Quoted_string (sprintf "ignore-%s" (Outputs.to_string outputs))
```
Comment (diml) [inline src/action.ml]: This should stay an atom

**Comment by diml** [comment]
BTW, we have an intern at Jane Street who is going to work on an ast/parser/printer generator for S-expressions from a BNF grammar, with support for completion in emacs, etc...

So we shouldn't worry too much about performances implication of validating strings too often for now.

File: src/action.ml (hunk 2 above)
Comment (Chris00) [reply src/action.ml]: So `outputs` is guaranteed to never break the Atom invariant?

**Comment by Chris00** [comment]
> These names are supposed to match OCaml types.

That makes sense if you use these types in the rest of the program.  If more private strings are used, maybe additional conversion functions should be added — unless you want to pass through strings because you want to re-check them.

I changed that and the `Quoted_string` above.

File: src/action.ml (hunk 2 above)
Comment (diml) [reply src/action.ml]: `Outputs.to_string outputs` is one of `"stdout"`, `"stderr"` or `"outputs"`

File: bin/main.ml [hunk 3]
```diff
@@ -872,7 +872,7 @@ let rules =
                      ; "targets", paths rule.targets ]
                    ; (match rule.context with
                       | None -> []
-                      | Some c -> ["context", Atom c.name])
+                      | Some c -> ["context", Sexp.atom c.name])
```
Comment (diml) [inline bin/main.ml]: That should be `atom_or_string`, context names can be (almost) any string

File: src/build_system.ml [hunk 4]
```diff
@@ -1085,7 +1085,8 @@ module Trace = struct
           Pmap.add acc ~key ~data)
         |> Path.Map.bindings
         |> List.map ~f:(fun (path, hash) ->
-          Sexp.List [ Atom (Path.to_string path); Atom (Digest.to_hex hash) ]))
+               Sexp.List [ Quoted_string (Path.to_string path);
```
Comment (diml) [inline src/build_system.ml]: This previous code is quite old, we can update it to use `Path.sexp_of_t`

File: src/sexp.ml [hunk 5]
```diff
@@ -5,9 +5,9 @@ include (Usexp : module type of struct include Usexp end
 
 let code_error message vars =
   code_errorf "%a" pp
-    (List (Atom message
+    (List (Quoted_string message
```
Comment (diml) [inline src/sexp.ml]: We should use `atom_or_string` here. Sometimes the message is just a function name, and it looks slightly nicer without the quotes IMO

File: src/sexp.ml [hunk 6]
```diff
@@ -5,9 +5,9 @@ include (Usexp : module type of struct include Usexp end
...
            :: List.map vars ~f:(fun (name, value) ->
-             List [Atom name; value])))
+             List [Atom (Atom.of_string name); value])))
```
Comment (diml) [inline src/sexp.ml]: The name is not guaranteed to be an atom. It is in general, but it can be whatever we think will be useful for debugging

**Comment by diml** [comment]
> That makes sense if you use these types in the rest of the program. If more private strings are used, maybe additional conversion functions should be added — unless you want to pass through strings because you want to re-check them.

What I would do is the following: for instance for module names, I would add a module `Module.Name` defining a type t in the same way you defined atoms. Then, since the invariant for module names is stronger than the invariant for atoms, I would define `Module.Name.sexp_of_t` as follow:

```ocaml
let sexp_of_t (T t) = Sexp.unsafe_atom_of_string t
```

> I changed that and the Quoted_string above.

Thanks

**Comment by Chris00** [comment]
@diml Your remarks have been answered.  I think that making more types private is for subsequent PRs.

**Pull request merged**